cql3/selection/selector.hh

```diff
@@ -109,7 +109,10 @@ public:
     virtual assignment_testable::test_result test_assignment(database& db, const sstring& keyspace, ::shared_ptr<column_specification> receiver) override {
         auto t1 = receiver->type->underlying_type();
         auto t2 = get_type()->underlying_type();
-        if (t1 == t2) {
+        // We want columns of `counter_type' to be served by `long_type' overloads with
+        // an `EXACT_MATCH'. Weak assignability between the two would lead to ambiguity
+        // because `WEAKLY_ASSIGNABLE' counter->blob conversion exists and would compete.
+        if (t1 == t2 || (t1 == long_type && t2->is_counter())) {
```

Comment:
An alternative solution is to create those aggregate functions for counter types. But I think it's not as good, because it won't work for user defined functions.

There is a risk here that this overload will compete with another overload that does have a counter type for the receiver, and we will lost it.

Reply:
Yet another option is to change test_result into a class that can express EXACT_MATCH, NOT_ASSIGNABLE, and WEAKLY_ASSIGNABLE(score), and code that checks for WEAKLY_ASSIGNABLE can select the best score. Not sure if there is such a place in the code base.

Reply:
@psarna @tgrabiec opinions please

Reply:
1. I think that instead of open coding `long_type`, `counter_cell_view::total_value_type()` can be used to show why it's ok for counters to use that type.
2. As for exact_match and weakly_assignable, @jul-stas I need more context first - why do counters fall back to `blob->blob` in the first place? Is it because they're weakly assignable to both blobs and long_type, but blobs just happen to be first on the list?

Reply:
@psarna 
1. I don't understand - do you mean unification of `long_type`'s and `counter_type`'s `native` types? Or... I don't know... a static assert on identity of both types?
2. Almost. Blob is weakly assignable to every type, per:
 https://github.com/jul-stas/scylla/blob/ccaaf74cb91dfd3623c3a4775fe80a922ed53508/types.cc#L2475
Now, there is no `sum(blob)->blob` nor `avg(blob)->blob`, but we do have `min(blob)->blob` and `max(blob)->blob`. There is no implicit conversion `counter->int/bigint`. So when called on counters, `sum`/`avg` respond "Invalid, none of its signatures match". But `min`/`max` called on counters find the only weakly assignable overload: `(blob)->blob` (as in the example in top message).
Now if I grant counters any additional weak assignability, say `counter->bigint`, then `sum`/`avg` will start to work correctly - by using the only weakly assignable overload: `sum(bigint)->bigint`. But `min`/`max` will stop working because of competing overloads: `blob` and `bigint` (and the message will be "ambiguous call" because, sadly, we have no notion of overloads' priorities). So if we don't have an exactly matching overload and we have >1 weakly matching overload => error.

Reply:
1. Oh, I just mean using `counter_cell_view::total_value_type()` instead of `long_type` in this very line :)
2. Ok, makes sense. It would be nice to have an assignability score and use that, but since we don't have it yet, I think that your way is simple and correct. Looks good.